veriphix/client.py

```diff
@@ -196,125 +261,176 @@ def refresh_randomness(self) -> None:
-        # StateVectorBackend because noiseless preparation
-        preparation_backend = StatevectorBackend()
-        preparation_backend.add_nodes(nodes=sorted(self.graph[0]), data=run.states)
-        self.blind_qubits(preparation_backend)
-
-        backend.add_nodes(nodes=sorted(self.graph[0]), data=preparation_backend.state)
-
-        tmp_measurement_db = self.measurement_db.copy()
-        # Modify the pattern to be all X-basis measurements, no shifts/signalling updates
-        # Warning should only work for BQP ie classical output
-        for node in self.measurement_db:
-            self.measurement_db[node] = graphix.command.M(node=node)
-
-        # TODO add measurements on output nodes?
-
-        sim = PatternSimulator(
-            backend=backend, pattern=self.clean_pattern, measure_method=self.measure_method, **kwargs
-        )
-        sim.run(input_state=None)
-
-        trap_outcomes = []
-        for trap in run.traps_list:
-            outcomes = [self.results[component] for component in trap]  # here
-            trap_outcome = sum(outcomes) % 2
-            trap_outcomes.append(trap_outcome)
-
-        self.measurement_db = tmp_measurement_db
-
-        return trap_outcomes
-
-    def delegate_pattern(self, backend: Backend, **kwargs) -> None:
-        self.prepare_states(backend)
-        self.blind_qubits(backend)
-        sim = PatternSimulator(
-            backend=backend, pattern=self.clean_pattern, measure_method=self.measure_method, **kwargs
-        )
-        sim.run(input_state=None)
-        self.decode_output_state(backend)
+            # 1 color = 1 test run = 1 collection of single-qubit traps
+            traps_list = [frozenset([colored_node]) for colored_node in nodes_by_color[color]]
+            traps: Traps = frozenset(traps_list)
+            test_run = TestRun(client=self, traps=traps)
+            test_runs.append(test_run)
+
+        # print(test_runs)
+        return test_runs
+
+    def sample_canvas(self):
+        N = self.trappifiedScheme.params.comp_rounds + self.trappifiedScheme.params.test_rounds
+        computation_rounds = set(random.sample(range(N), self.trappifiedScheme.params.comp_rounds))
+
+        rounds_run = dict()
+        for r in range(N):
+            if r in computation_rounds:
+                rounds_run[r] = self.computationRun
+            else:
+                rounds_run[r] = random.choice(self.test_runs)
+        return rounds_run
+
+    def delegate_canvas(self, canvas: dict, backend: Backend, **kwargs):
+        outcomes = dict()
+        for r in canvas:
+            round_outcome = canvas[r].delegate(backend=backend, **kwargs)
+            if round_outcome == {}:
+                outcomes[r] = backend.state
+            else:
+                outcomes[r] = round_outcome
+            # Ugly reset of backend. Needed in case of quantum output
+            # TODO: how to do that cleaner ?
```

Comment:
Since you refresh the backend at each iteration, you don't need a preinitialized backend in the first place, so it would be cleaner for `delegate_canvas` to take a `Backend` class instead of an instance and be responsible for instantiating a fresh backend at each round.